things/postgres/groups.go

```diff
@@ -0,0 +1,631 @@
+		  			  %s ORDER BY id LIMIT :limit OFFSET :offset;`, mq)
+
+	params := map[string]interface{}{
+		"userID":   userID,
+		"limit":    limit,
+		"offset":   offset,
+		"metadata": m,
+	}
+
+	rows, err := gr.db.NamedQueryContext(ctx, q, params)
+	if err != nil {
+		return groups.GroupPage{}, errors.Wrap(errSelectDb, err)
+	}
+	defer rows.Close()
+
+	var items []groups.Group
+	for rows.Next() {
+		dbgr := dbGroup{}
+		if err := rows.StructScan(&dbgr); err != nil {
+			return groups.GroupPage{}, errors.Wrap(errSelectDb, err)
+		}
+		gr := toGroup(dbgr)
+		if err != nil {
+			return groups.GroupPage{}, err
+		}
+		items = append(items, gr)
+	}
+
+	cq := fmt.Sprintf(`SELECT COUNT(*) 
+					   FROM thing_group_relations gr, thing_groups g
+					   WHERE gr.group_id = g.id and gr.thing_id = :userID %s;`, mq)
+
+	total, err := total(ctx, gr.db, cq, params)
+	if err != nil {
+		return groups.GroupPage{}, errors.Wrap(errSelectDb, err)
+	}
+
+	page := groups.GroupPage{
+		Groups: items,
+		PageMetadata: groups.PageMetadata{
+			Total:  total,
+			Offset: offset,
+			Limit:  limit,
+		},
+	}
+
+	return page, nil
+}
+
+func (gr groupRepository) Assign(ctx context.Context, thingID, groupID string) error {
+	dbr, err := toDBGroupRelation(thingID, groupID)
+	if err != nil {
+		return errors.Wrap(groups.ErrAssignToGroup, err)
+	}
+
+	qIns := `INSERT INTO thing_group_relations (group_id, thing_id) VALUES (:group_id, :thing_id)`
+	_, err = gr.db.NamedQueryContext(ctx, qIns, dbr)
+	if err != nil {
+		pqErr, ok := err.(*pq.Error)
+		if ok {
+			switch pqErr.Code.Name() {
+			case errInvalid, errTruncation:
+				return errors.Wrap(groups.ErrMalformedEntity, err)
+			case errDuplicate:
+				return errors.Wrap(groups.ErrGroupConflict, err)
+			case errFK:
+				return errors.Wrap(groups.ErrNotFound, err)
+			}
+		}
+		return errors.Wrap(groups.ErrAssignToGroup, err)
+	}
+
+	return nil
+}
+
+func (gr groupRepository) Unassign(ctx context.Context, userID, groupID string) error {
+	q := `DELETE FROM thing_group_relations WHERE thing_id = :thing_id AND group_id = :group_id`
+	dbr, err := toDBGroupRelation(userID, groupID)
+	if err != nil {
+		return errors.Wrap(groups.ErrNotFound, err)
+	}
+	if _, err := gr.db.NamedExecContext(ctx, q, dbr); err != nil {
+		return errors.Wrap(groups.ErrGroupConflict, err)
+	}
+	return nil
+}
+
+type dbGroup struct {
+	ID          string        `db:"id"`
+	Name        string        `db:"name"`
+	OwnerID     uuid.NullUUID `db:"owner_id"`
+	ParentID    uuid.NullUUID `db:"parent_id"`
+	Description string        `db:"description"`
+	Metadata    dbMetadata    `db:"metadata"`
+	Path        string        `db:"path"`
+	Level       int           `db:"level"`
+}
+
+type dbGroupPage struct {
+	ID       uuid.NullUUID `db:"id"`
+	OwnerID  uuid.NullUUID `db:"owner_id"`
+	ParentID uuid.NullUUID `db:"parent_id"`
+	Metadata dbMetadata    `db:"metadata"`
+	Limit    uint64
+	Offset   uint64
+	Size     uint64
+}
+
+func toUUID(id string) (uuid.NullUUID, error) {
```

Comment:
This looks way too complicated. Can you use `uuid.FromStringOrNil(id)` or at least simplify this method?

Reply:
This whole Null UUID handling looks complicated. Can you try to simplify it?

Reply:
I already did simplification and this best I can do